🐛 Respect template HTML for strings of many elements

caroqliu opened the pull request:
`amp-mustache` `setHtml` wraps an HTML string of multiple elements with an outer div, silently turning
```
"<div>A</div><div>B</div><div>C</div>"
```
into
```html
<div>
  <div>A</div>
  <div>B</div>
  <div>C</div>
</div>
```

This PR allows `setHtml` to return many elements as an array instead of adding an outer `<div>`, which is not well documented behavior. In the above example it will now return:
```html
[ <div>A</div>,  <div>B</div>,  <div>C</div> ]
```

Notes:
- The result of passing singular elements to `setHtml` (`<div>A</div>`) remains the same (`<div>A</div>`).
- Setting non-element HTML (`"abc"`) results in `<div>abc</div>` prior to and after this change.
- Setting an HTML string that consists of both elements and text nodes will wrap text nodes in `<div>` elements prior to the return. That is, `"<div>A</div>B<div>C</div>"` returns `[ <div>A</div>, <div>B</div>, <div>C</div> ]`. Alternatively we could skip by returning `[ <div>A</div>, <div>C</div> ]` or default to the prior behavior by returning:
  ```
  <div>
    <div>A</div>
    B
    <div>C</div>
  </div>
  ```
  None of these options seem that intuitive so will defer to folks who may have stronger opinions here.

cc @choumx

choumx commented:
> Setting non-element HTML ("abc") results in <div>undefined</div> both prior to and after this change.

Shouldn't this return `<div>abc</div>`?

choumx commented on src/service/template-impl.js:
```diff
@@ -119,6 +119,33 @@ export class BaseTemplate {
+      } else if (n.nodeType == /* ELEMENT */ 1) {
+        elements.push(dev().assertElement(n));
+      }
+    }
```
Can we call `unwrap()` internally instead?

caroqliu replied:
Could you elaborate on the idea? Instead of what?

choumx replied:
Instead of duplicating the code.

caroqliu replied:
Are you asking for `forceUnwrap` to use `unwrap` on each of its root's child nodes? Or are you asking for `setHtml` to `unwrap` instead of `forceUnwrap` without any modifications to `unwrap`?

caroqliu replied:
To decrease duplicated code, I had `forceUnwrap` call `unwrap` instead of the other way around. LMK if you had something else in mind.

caroqliu replied:
Yes, `setHtml` wraps once prior to the purification-wrapping to guarantee the post-purification HTML can be unwrapped to a single element, then it will `unwrapChildren` for the target contents.

So the call order for amp-mustache 0.2 would be:
- `setHtml("abc")`
  - `purifyAndSetHtml_("<div>abc</div>")`
    - `purifyHtml("<div><div>abc</div></div>")`
    - `tryUnwrap(<div><div>abc</div></div>)`
  - `unwrapChildren(<div>abc</div>)`
- finds `abc` as first child, returns it wrapped `<div>abc</div>`

Another example: 
- `setHtml("<a>abc</a>")`
  - `purifyAndSetHtml_("<div><a>abc</a></div>")`
    - `purifyHtml("<div><div><a>abc</a></div></div>")`
    - `tryUnwrap(<div><div><a>abc</a></div></div>)`
  - `unwrapChildren(<div><a>abc</a></div>)`
- finds `<a>abc</a>` as first child, returns `<a>abc</a>`

I see your point in the first example that the `<div>` passed to `unwrapChildren` isn't the same as the `<div>` that gets returned. In that case I added a guard for the child traversal:
```js
if (!root.firstElementChild) {
  return root;
}
```

This can also prevent `unwrapChildren(<a>abc</a>)`  from returning `<div>abc</div>`, which might not be expected at a glance. This edge case isn't particularly applicable to the way it is used now, but it's worth covering in case it is used this way in the future.

choumx replied:
Thanks for iterating on this given vague direction. IMO still pretty confusing to read... what if we created an iterator helper instead that both functions call?

```js
function unwrap_(root, callback) {
	for (let n = root.firstChild; n != null; n = n.nextSibling) {
	  if (n.nodeType == /* TEXT */ 3) {
	    const text = n.textContent.trim();
	    callback(text);
	  } else if (n.nodeType == /* COMMENT */ 8) {
	    // Ignore comments.
	  } else if (n.nodeType == /* ELEMENT */ 1) {
	    callback(dev().assertElement(n));
	  }
	}
}

tryUnwrap(root) {
	let onlyChild;
	this.unwrap_(root, c => {
		if (onlyChild === undefined && c.nodeType) {
			onlyChild = c;
		} else {
      onlyChild = null;
    }
	});
	return onlyChild;
}

unwrapChildren(root) {
	let children = [];
	this.unwrap_(root, c => {
		if (typeof c == 'string') {
			// Wrap in <div> and push.
		} else {
			children.push(c);
		}
	});
  return children;
}
```

caroqliu replied:
Applied your recommendation with the following differences:
- `unwrap_` is renamed to `visitChildren_`
- `callback` is only applied to text nodes when non-empty
- `unwrapChildren` still guards child traversal and returns type `Element` if it only finds one child
- `tryUnwrap` returns `onlyChild || root`

Question: Is there a reason to check `nodeType` in `tryUnwrap` and `typeof` in `unwrapChildren`? Why don't they distinguish their arguments in the same way?

caroqliu replied:
Also, in this case are we increasing code size in favor of readabilit

caroqliu commented:
> > Setting non-element HTML ("abc") results in undefined both prior to and after this change.
> 
> Shouldn't this return `<div>abc</div>`?

Yeah, updated the description and added a test case.

choumx commented on src/service/template-impl.js:
```diff
@@ -94,29 +94,35 @@ export class BaseTemplate {
+   * @return {!Element|!Array<!Element>}
+   * @protected @final
+   */
+  forceUnwrap(root) {
```
Nit: Maybe `tryUnwrapChild` and `unwrapChildren` are clearer.

choumx commented on src/service/template-impl.js:
```diff
@@ -94,29 +94,35 @@ export class BaseTemplate {
   unwrap(root) {
-    let singleElement = null;
+    const unwrapped = this.forceUnwrap(root);
+    return dev().assertElement(unwrapped.length ? root : unwrapped);
```
`isArray(unwrapped) ? root : unwrapped`

caroqliu replied:
We also want to return `root` if `unwrapped` is `[]`, no?

caroqliu commented on src/service/template-impl.js:
```diff
@@ -86,37 +86,70 @@ export class BaseTemplate {
+   */
+  unwrapChildren(root) {
+    const children = [];
+    if (!root.firstElementChild) {
```
I double-checked the `<p>text</p>` case and `firstElementChild` is also `null`, so it would return `<p>text</p>`. I would consider it unexpected for `<p>text</p>` to return `<div>text</div>`, but because nothing relies on this behaving one way or the other, I don't feel too strongly either way. Removed.

caroqliu commented on src/service/template-impl.js:
```diff
@@ -86,37 +86,70 @@ export class BaseTemplate {
+        children.push(c);
+      }
+    });
+    return children.length === 1 ? children[0] : children;
```
This makes a difference to `amp-form`, but I can handle it there instead.

caroqliu commented on src/service/template-impl.js:
```diff
@@ -86,37 +86,70 @@ export class BaseTemplate {
+  /**
+   * Helps the template implementation to unwrap the root element. The root
+   * element can be unwrapped only when it contains a single element or a
+   * single element surrounded by empty text nodes.
```
Updated the comment. Leaving `|| root` in the return for now.

caroqliu commented on extensions/amp-form/0.1/amp-form.js:
```diff
@@ -1296,12 +1296,19 @@ export class AmpForm {
+              renderContainer = rendered;
+            }
+            renderContainer.id = messageId;
+            renderContainer.setAttribute('i-amphtml-rendered', '');
```
Done.

choumx commented on extensions/amp-form/0.1/amp-form.js:
```diff
@@ -1296,12 +1296,24 @@ export class AmpForm {
-            rendered.setAttribute('i-amphtml-rendered', '');
+            // TODO(caroqliu): Simplify section appending rendered contents to DOM.
+            let renderContainer;
+            if (isArray(rendered)) {
```
Thinking about it now, would we ever actually expect an array for form renders? I.e. is input `data` ever an array?

caroqliu replied:
Omg, no! 🤦 
https://github.com/ampproject/amphtml/blob/98d57b0ebb28b82b9c32b46e4a400d395070ecfd/extensions/amp-form/0.1/amp-form.js#L1281-L1287

I'll replace the changes on amp-form with a `devAssert` that the result is not an array.

choumx replied:
Yea, we're handling both array and non-array cases in a single codepath for SSR. But I still think the same invariant applies -- we should only receive an array of elements if `data` is an array.

Since `setHtml` will always return an array now, you could just assert that length == 1. 

Anyways, let's leave this for the "simplify" TODO. You can also file an issue and reference this PR thread instead of self-assigning.

choumx commented on src/service/template-impl.js:
```diff
@@ -86,37 +86,64 @@ export class BaseTemplate {
+   * Unwraps the root element and returns any children in an array.
+   * Text node children are normalized inside a <div>.
+   * @param {!Element} root
+   * @return {!Element|!Array<!Element>}
```
Remove `!Element`.
